File: leetcode-cn/0009.0_Palindrome_Number.py
'''
approach: math
Time: O(logN)
Space: O(1)

执行用时：84 ms, 在所有 Python3 提交中击败了35.05% 的用户
内存消耗：14.8 MB, 在所有 Python3 提交中击败了76.76% 的用户
'''

class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x < 0:
            return False

        x_bak = x
        reverse = 0
        while x:
            reverse *= 10
            reverse += (x % 10)
            x //= 10
            
        return x_bak == reverse



'''
执行用时：80 ms, 在所有 Python3 提交中击败了46.20% 的用户
内存消耗：14.9 MB, 在所有 Python3 提交中击败了30.70% 的用户
'''

class Solution:
    def isPalindrome(self, x: int) -> bool:
        return str(x) == str(x)[:: -1]

File: csharp/0009.0_Palindrome_Number.cs
/*
执行用时：36 ms, 在所有 C# 提交中击败了83.69% 的用户
内存消耗：28.6 MB, 在所有 C# 提交中击败了97.02% 的用户
通过测试用例：11510 / 11510
*/
public class Solution {
    public bool IsPalindrome(int x) {
        if (x < 0) return false;
        int back = 0, xx = x;
        while (x != 0) {
            back = back * 10 + x % 10;
            x /= 10;
        }
        return back == xx;
    }
}

/*
执行用时：44 ms, 在所有 C# 提交中击败了46.22% 的用户
内存消耗：28.6 MB, 在所有 C# 提交中击败了95.76% 的用户
通过测试用例：11510 / 11510
*/
public class Solution {
    public bool IsPalindrome(int x) {
        if (x < 0 || (x % 10 == 0 && x != 0)) return false;
        int revertedNumber = 0;
        while (x > revertedNumber) {
            revertedNumber = revertedNumber * 10 + x % 10;
            x /= 10;
        }
        return x == revertedNumber || x == revertedNumber / 10;
    }
}